feat(build): implement dynamic versioning with hatch-vcs

camjac251 opened the pull request:
Implements dynamic versioning using hatch-vcs to automatically derive version numbers from git tags, eliminating the need for manual version management in source files.

This change moves version control to git tags, where hatch-vcs automatically generates version strings during builds. Development builds include commit hashes and dates, while release builds use clean semantic versions from tags.

## Configuration

The build system now uses hatch-vcs in pyproject.toml:

```toml
[build-system]
requires = ["hatchling", "hatch-vcs"]

[project]
dynamic = ["version"]

[tool.hatch.version]
source = "vcs"
```

## Version Formats

- **Release builds**: 4.0.1 (from git tag v4.0.1)
- **Development builds**: 4.0.2.dev1+g9e71322e6.d20251112 (includes commit hash and date)

## Creating New Versions

The update_version.py script now creates git tags:

```bash
# Create specific version
uv run scripts/update_version.py 4.2.0

# Or bump semantically
uv run scripts/update_version.py --bump major
uv run scripts/update_version.py --bump minor
uv run scripts/update_version.py --bump patch
```

## Implementation Notes

The version.py module now imports from the generated _version.py file created by hatch-vcs during builds, with fallbacks to importlib.metadata for editable installs. The generated _version.py is added to .gitignore since it's build-time generated.

This follows Python packaging best practices and matches how major projects like setuptools-scm manage versions. The approach provides a single source of truth (git tags) while maintaining compatibility with all installation methods.

ofriw commented:
@camjac251  - Thank you for this excellent work on dynamic versioning! The hatch-vcs implementation is exactly what we need. 🎉

I'd like to enhance this PR to support our upcoming beta channel launch. This requires updating the version validation regex to accept PEP 440 pre-release formats (`4.1.0b1`, `4.1.0rc2`, etc.).

**Two options - your choice:**

**Option A (Faster):** I push the enhancements directly to your branch. You review my commits, and we merge with full credit to you for the feature.

**Option B (Learning opportunity):** I document the exact changes needed (regex pattern, docs, tests), and you implement them. Great way to learn PEP 440 versioning!

Either way works for me - what's your preference? If I don't hear back within 48 hours, I'll go with Option A to keep momentum, but you'll still review before merge.

P.S. This is **adding scope**, not fixing issues - your implementation is solid!

camjac251 commented:
@ofriw I went with Option B and implemented PEP 440 pre-release support for the beta channel.

**Changes:**
- Updated version validation to accept pre-release formats: `4.1.0a1`, `4.1.0b2`, `4.1.0rc3`, plus alternative spellings like `alpha`, `beta`, `preview`
- Rewrote `update_version.py` to use argparse instead of manual sys.argv parsing - added proper `--help`, type hints, and cleaner error handling
- Added pre-release support to the bump command: `--bump minor b1` creates beta versions
- Updated CLAUDE.md and CONTRIBUTING.md with examples and workflow documentation

**Usage:**
```bash
# Beta releases
uv run scripts/update_version.py 4.1.0b1
uv run scripts/update_version.py --bump minor b1

# Release candidates
uv run scripts/update_version.py 4.1.0rc1
```

Should be ready for review.
